src/material/core/datetime/date-selection.ts

```diff
@@ -0,0 +1,159 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DateAdapter} from '@angular/material/core';
+import {Subject} from 'rxjs';
+
+export abstract class MatDateSelection<D> {
+  valueChanges = new Subject<void>();
+
+  constructor(protected readonly adapter: DateAdapter<D>) {}
+
+  dispose() {
+    this.valueChanges.complete();
+  }
+
+  abstract add(date: D | null): void;
+  abstract clone(): MatDateSelection<D>;
+  abstract getFirstSelectedDate(): D|null;
```

Comment:
Same for `getFirstSelectedDate` and `getLastSelectedDate`, I'm not 100% sure if we're going to need these methods, so lets remove them and add them if we do